drivers/gles3/rasterizer_scene_gles3.cpp

```diff
@@ -2015,6 +2025,87 @@ void RasterizerSceneGLES3::_render_list(RenderList::Element **p_elements, int p_
 			skeleton = storage->skeleton_owner.getornull(e->instance->skeleton);
 		}
 
+		bool started_query = false;
+
+		bool use_oq = false;
+
+		if (e->instance->base_type == VS::INSTANCE_MESH) {
+			//The overhead makes OQ counterproductive for too simple meshes
+			RasterizerStorageGLES3::Surface *s = static_cast<RasterizerStorageGLES3::Surface *>(e->geometry);
+			use_oq = p_occlusion_queries && s->array_len > 256;
+		}
+		if (use_oq) {
+			if (!occlusion_queries->has(e->instance->get_id())) {
+				// I am honestly unsure about this part. I would prefer to create enough queries in advance, store them in a list and then just get "new" queries from there instead.
+				// However, I found other parts of the codebase that create stuff only just when it is needed.
+				// For now I am gonna leave this here, I will probably need to address this after someone has reviewed this PR. I would also like feedback on whether I should move this to storage (or an inner class of storage)
+				GLuint query = 0;
+				glGenQueries(1, &query);
+
+				occlusion_queries->insert(e->instance->get_id(), OcclusionQueryData{ OcclusionQueryPrevAction::NoAction, 0, query });
+			}
+
+			OcclusionQueryData &data = occlusion_queries->find(e->instance->get_id())->value();
+			uint64_t cur_frame = storage->frame.count;
+			GLuint query_result = 0;
+
+			bool query_already_started = false;
+
+			if (data.prev_action != OcclusionQueryPrevAction::NoAction) {
+				if (data.prev_frame == cur_frame) {
+					// we already started a OQ for this object in this frame (e.g. in the early depth pass)
+					// just repeat what we did before and don't start a new query
+					if (data.prev_action == OcclusionQueryPrevAction::Culled) {
+						//since we already have a query there is no reason to draw an AABB
+						continue;
+					}
+
+					//skip starting the query and just render the mesh
+					query_already_started = true;
+				} else {
+					GLuint query_result_available = 0;
+					glGetQueryObjectuiv(data.query, GL_QUERY_RESULT_AVAILABLE, &query_result_available);
+					if (query_result_available == GL_FALSE) {
+						query_result = 1; //let's just assume this is visible
+					} else {
+						glGetQueryObjectuiv(data.query, GL_QUERY_RESULT, &query_result);
+					}
```

Comment:
I wonder if you check if ``cur_frame - prev_frame >= 2`` here to allow for up to 2 frames in flight. 

That being said, the description of GL_QUERY_RESULT_AVAILABLE makes it sound like it doesn't have to wait for the GPU to catch up before returning (as the function would be pointless if it did)